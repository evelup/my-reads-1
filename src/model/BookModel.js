class BookModel {
  constructor({
    id, title = '', subtitle = '', authors = [], imageLinks = {}, shelf = 'none',
  }) {
    this.id = id;
    this.title = title;
    this.subtitle = subtitle;
    this.authors = authors;
    this.imageLinks = imageLinks;
    this.shelf = shelf;
  }
}

export default BookModel;
