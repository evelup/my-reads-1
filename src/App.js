import React from 'react';
import _ from 'underscore';
import { Route } from 'react-router-dom';
import * as BooksAPI from './BooksAPI'
import './App.css';
import SearchPage from './components/SearchPage';
import HomePage from './components/HomePage';
import {Book} from './model/Book';

class BooksApp extends React.Component {

  state = {
    books: {}
  };

  constructor() {
    super();
    this.handleShelfUpdate = this.handleShelfUpdate.bind(this);
  }

  componentDidMount = () => {
    BooksAPI.getAll().then((response) => {
      // this.setState( { books: response } );
      // TODO: Extract logic to BookService
      this.setState(state => ({
        books: _.groupBy(response.map(_book => new Book(_book)), 'shelf')
      }));
    });
  }

  handleShelfUpdate(book) {
    console.log(book);
    BooksAPI.update(book, book.shelf).then(response => {
      // TODO: Update state.books
      console.log(response);
    });
  }

  render() {
    return (
      <div className="app">
      <Route
          exact
          path="/"
          render={() => (<HomePage
            books={this.state.books}
            onUpdateBook={this.handleShelfUpdate}
          />)}
        />
        <Route
          path="/search"
          render={() => (<SearchPage
            onUpdateBook={this.handleShelfUpdate}
          />)}
        />
      </div>
    )
  }
}

export default BooksApp
