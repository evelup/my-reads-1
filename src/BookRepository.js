import BookModel from './model/BookModel';
import * as BooksAPI from './BooksAPI';

const MAX_API_RESULTS = '20';
export const getAllBooks = () =>
  new Promise((resolve, reject) => {
    BooksAPI.getAll()
      .then(response => resolve(response.map(book => new BookModel(book))))
      .catch((error) => {
        reject(error);
      });
  });

export const update = book =>
  new Promise(
    (resolve) => {
      BooksAPI.update(book, book.shelf).then((response) => {
        resolve(book);
      });
    },
    (error) => {
      console.log(error);
    },
  );

export const search = (query, allBooks) =>
  new Promise((resolve, reject) => {
    BooksAPI.search(query, MAX_API_RESULTS).then((results) => {
      if (results.error) {
        reject(new Error(results.error));
      } else {
        const books = results.map(book => new BookModel(book));
        books.forEach((book, index) => {
          allBooks.forEach((bookOnTheShelf) => {
            if (book.id === bookOnTheShelf.id) {
              books[index].shelf = bookOnTheShelf.shelf;
            }
          });
        });
        resolve(books);
      }
    });
  });
