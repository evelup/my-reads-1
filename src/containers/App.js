import React from 'react';
import { Route } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.min.css';
import './App.css';
import SearchPage from '../components/SearchPage';
import HomePage from '../components/HomePage';
import * as BookRepository from '../BookRepository';

class BooksApp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      allBooks: [],
      searchResultBooks: [],
      hasError: false,
      isWaitingResponse: false,
    };
    this.handleShelfUpdate = this.handleShelfUpdate.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
    this.clearSearch = this.clearSearch.bind(this);
  }

  componentDidMount = () => {
    this.setState({ isWaitingResponse: true });
    BookRepository.getAllBooks()
      .then((allBooks) => {
        this.setState({ allBooks, isWaitingResponse: false });
      })
      .catch((error) => {
        this.setState({ hasError: true, isWaitingResponse: false });
      });
  };

  handleShelfUpdate(book) {
    // this.setState({ isWaitingResponse: true });
    BookRepository.update(book)
      .then((updatedBook) => {
        this.setState(prevState => ({
          allBooks: prevState.allBooks.filter(b => b.id !== book.id).concat([updatedBook]),
          isWaitingResponse: false,
        }));
      })
      .then(() => {
        toast.info(`The book: ${book.title} was moved to ${book.shelf}`, {
          position: toast.POSITION.BOTTOM_CENTER,
        });
      });
  }

  handleSearch(query) {
    this.setState({ isWaitingResponse: true });
    BookRepository.search(query, this.state.allBooks)
      .then((searchResultBooks) => {
        this.setState({ searchResultBooks, hasError: false, isWaitingResponse: false });
      })
      .catch((error) => {
        this.setState({ searchResultBooks: [], hasError: true, isWaitingResponse: false });
      });
  }

  clearSearch() {
    this.setState({ searchResultBooks: [], hasError: false });
  }

  render() {
    return (
      <div className="app">
        <ToastContainer />
        <Route
          exact
          path="/"
          render={() => (
            <HomePage
              allBooks={this.state.allBooks}
              isWaitingResponse={this.state.isWaitingResponse}
              onUpdateBook={this.handleShelfUpdate}
            />
          )}
        />
        <Route
          path="/search"
          render={() => (
            <SearchPage
              searchResultBooks={this.state.searchResultBooks}
              isWaitingResponse={this.state.isWaitingResponse}
              hasError={this.state.hasError}
              onSearch={this.handleSearch}
              onUpdateBook={this.handleShelfUpdate}
              onClearSearch={this.clearSearch}
            />
          )}
        />
      </div>
    );
  }
}

export default BooksApp;
