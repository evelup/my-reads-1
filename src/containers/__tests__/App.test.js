import React from 'react';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import App from '../App';
import { shallow, mount } from 'enzyme';
import * as BookRepository from '../../BookRepository';
import allBooksResponse from './allbooks.json';
import BookModel from '../../model/BookModel';
import { MemoryRouter as Router, withRouter } from 'react-router-dom'

jest.mock('../../BookRepository');

beforeEach(() => {
  BookRepository.getAllBooks = jest.fn(() =>
    Promise.resolve([].concat(...allBooksResponse).map(book => new BookModel(book))));
});

it('renders without crashing', () => {
  const tree = renderer.create(<Router><App /></Router>).toJSON();
  expect(tree).toMatchSnapshot();
});
