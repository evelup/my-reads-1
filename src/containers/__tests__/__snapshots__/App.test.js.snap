// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="app"
>
  <div />
  <div
    className="list-books"
  >
    <div
      className="list-books-title"
    >
      <h1>
        MyReads
      </h1>
    </div>
    <div
      className="list-books-content"
    >
      <div>
        <div
          className="bookshelf"
        >
          <h2
            className="bookshelf-title"
          >
            Currently Reading
          </h2>
          <div
            className="align-center"
          >
            <div>
              <div
                className="css-1ucfsrd"
              />
              <div
                className="css-rho67t"
              />
              <div
                className="css-1ucfsrd"
              />
            </div>
          </div>
        </div>
        <div
          className="bookshelf"
        >
          <h2
            className="bookshelf-title"
          >
            Want to Read
          </h2>
          <div
            className="align-center"
          >
            <div>
              <div
                className="css-1ucfsrd"
              />
              <div
                className="css-rho67t"
              />
              <div
                className="css-1ucfsrd"
              />
            </div>
          </div>
        </div>
        <div
          className="bookshelf"
        >
          <h2
            className="bookshelf-title"
          >
            Read
          </h2>
          <div
            className="align-center"
          >
            <div>
              <div
                className="css-1ucfsrd"
              />
              <div
                className="css-rho67t"
              />
              <div
                className="css-1ucfsrd"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="open-search"
    >
      <a
        href="/search"
        onClick={[Function]}
      >
        Add a book
      </a>
    </div>
  </div>
</div>
`;
